File: v1/remotix-rust/src/sentinella.rs
//! Chi guarda se c'e' qualcuno **davanti alla macchina**.
//!
//! # La regola, e perche' esiste
//!
//! Una sola sessione grafica per utente (SPECIFICA.md §3.4). Se l'utente ha
//! gia' una sessione grafica **locale**, la connessione RDP si rifiuta; se la
//! sessione locale compare mentre l'RDP e' in corso, **la locale vince** e
//! l'RDP cade.
//!
//! Non e' una regola di comodo: due sessioni grafiche dello stesso utente sullo
//! stesso `$XDG_RUNTIME_DIR` si contendono il gestore systemd dell'utente, il
//! bus di sessione, l'agente delle chiavi e i portali — cioe' proprio le cose
//! che nella fase 5 si e' faticato a far funzionare una volta sola. Il guasto
//! che ne segue non si presenta come «due sessioni»: si presenta come
//! applicazioni che non si aprono e impostazioni che tornano indietro.
//!
//! Le sessioni **testuali** convivono liberamente, ed e' importante che sia
//! cosi': REMOTIX stesso oggi gira dentro una sessione SSH.
//!
//! # Cosa conta come «grafica locale»
//!
//! Quattro condizioni insieme, lette da `systemd-logind`:
//!
//! | proprieta' | valore | perche' |
//! |---|---|---|
//! | `User`  | il nostro uid   | le sessioni di altri utenti non ci riguardano |
//! | `Seat`  | non vuoto       | un seat e' hardware vero: schermo e tastiera attaccati |
//! | `Type`  | `wayland`/`x11`/`mir` | `tty` e' testuale, e deve poter convivere |
//! | `Class` | `user`          | esclude `greeter`, `lock-screen`, `manager`, `background` |
//!
//! `Remote` deve inoltre essere falso: una sessione X11 inoltrata da lontano
//! non e' qualcuno davanti alla macchina.
//!
//! Misurato nella VM il 3 agosto: la sessione SSH in cui girano REMOTIX **e la
//! sessione GNOME che REMOTIX avvia** e' una sola, `Type=tty`, senza seat.
//! `gnome-session` non ne crea una propria — sta nel cgroup di chi lo ha
//! avviato — quindi la sessione remota non si conta da sola. Per prudenza la si
//! esclude comunque per identificatore: se un domani REMOTIX venisse avviato
//! dentro una sessione registrata come grafica, senza quell'esclusione si
//! sfratterebbe da solo, e il sintomo — «rifiuta sempre tutti» — non farebbe
//! sospettare la causa.
//!
//! # Perche' un segnale **e** un ripasso
//!
//! §3.4 lo dice: interrogare logind alla connessione non basta, perche' la
//! sessione locale puo' comparire dopo. Ci si sottoscrive quindi a
//! `SessionNew`/`SessionRemoved`.
//!
//! Il ripasso periodico non e' ridondanza inutile: il `Type` di una sessione
//! **cambia dopo la nascita** — `gdm` la registra e poi la promuove — quindi un
//! `SessionNew` visto troppo presto la mostrerebbe ancora testuale. Ogni due
//! secondi si riguarda; e' una manciata di chiamate D-Bus al minuto.
//!
//! # Se logind non c'e'
//!
//! Si prosegue **senza** la regola, dichiarandolo nel registro. E' una scelta:
//! l'alternativa — rifiutare tutti — trasformerebbe un bus non raggiungibile in
//! un server inaccessibile senza spiegazione. Chi non ha logind non ha nemmeno
//! il modo di aprire la sessione locale che stiamo temendo.

use core::time::Duration;
use std::sync::Arc;

use anyhow::{Context as _, Result};
use futures_util::StreamExt as _;
use tokio::sync::watch;
use tracing::{debug, info, warn};
use zbus::zvariant::OwnedObjectPath;
use zbus::{Connection, Proxy};

/// Ogni quanto si riguarda l'elenco anche senza segnali.
const RIPASSO: Duration = Duration::from_secs(2);

/// I tipi di sessione che occupano lo schermo.
const TIPI_GRAFICI: [&str; 3] = ["wayland", "x11", "mir"];

/// Una sessione grafica locale trovata, nella forma in cui si racconta.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct SessioneLocale {
    pub id: String,
    pub tipo: String,
    pub seat: String,
}

impl core::fmt::Display for SessioneLocale {
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        write!(f, "sessione {} ({} su {})", self.id, self.tipo, self.seat)
    }
}

/// Sorveglia la comparsa e la scomparsa della sessione grafica locale.
///
/// Chi la interroga non fa I/O: lo stato lo tiene aggiornato un compito a
/// parte. Serve perche' la risposta va data **nell'istante** in cui arriva una
/// connessione — interrogare D-Bus li' significherebbe far aspettare il client
/// per una risposta che quasi sempre e' «no».
pub struct Sentinella {
    stato: watch::Sender<Option<SessioneLocale>>,
}

impl Sentinella {
    /// Prepara la sentinella e fa **subito** il primo controllo.
    ///
    /// Il primo controllo e' atteso di proposito: se la macchina ha gia' una
    /// sessione grafica locale all'avvio di REMOTIX, non deve esistere una
    /// finestra iniziale in cui si entra lo stesso.
    pub async fn avvia() -> Arc<Self> {
        let (stato, _) = watch::channel(None);
        let sentinella = Arc::new(Self { stato });

        match Logind::apri().await {
            Ok(logind) => {
                let iniziale = logind.cerca().await;
                match &iniziale {
                    Some(sessione) => warn!(%sessione, "c'e' gia' una sessione grafica locale: le connessioni saranno rifiutate"),
                    None => info!("nessuna sessione grafica locale: si puo' entrare"),
                }
                sentinella.stato.send_replace(iniziale);
                tokio::spawn(sorveglia(Arc::clone(&sentinella), logind));
            }
            Err(errore) => {
                // Non si nasconde dietro un `debug`: significa che una delle
                // regole di §3.4 non e' in vigore, e chi legge il registro deve
                // poterlo sapere senza andarlo a cercare.
                warn!(
                    errore = %format!("{errore:#}"),
                    "logind non raggiungibile: la regola della sessione locale non e' applicata"
                );
            }
        }
        sentinella
    }

    /// La sessione grafica locale in corso, se c'e'.
    pub fn presente(&self) -> Option<SessioneLocale> {
        self.stato.borrow().clone()
    }

    /// Si risolve quando c'e' una sessione grafica locale — subito, se c'e' gia'.
    ///
    /// # Sicurezza rispetto all'annullamento
    ///
    /// Vive dentro la `select!` che serve la connessione, quindi va potuto
    /// abbandonare a meta' senza conseguenze: un `watch::Receiver` non consuma
    /// nulla quando lo si molla, e il valore resta li' per chi lo riprende.
    ///
    /// Se il compito di sorveglianza non c'e' — logind assente — questo futuro
    /// non si risolve mai, che e' esattamente il comportamento voluto: la
    /// connessione prosegue senza la regola.
    pub async fn attendi_comparsa(&self) -> SessioneLocale {
        let mut ricevitore = self.stato.subscribe();
        loop {
            if let Some(sessione) = ricevitore.borrow_and_update().clone() {
                return sessione;
            }
            if ricevitore.changed().await.is_err() {
                core::future::pending::<()>().await;
            }
        }
    }

    /// Si risolve quando la sessione grafica locale non c'e' piu' — subito, se
    /// non c'era.
    ///
    /// Serve a chi agisce sulla comparsa: senza un modo di aspettare che finisca
    /// ripeterebbe l'azione a ogni giro, perche' «c'e' una sessione locale»
    /// resta vero finche' quella dura.
    pub async fn attendi_scomparsa(&self) {
        let mut ricevitore = self.stato.subscribe();
        loop {
            if ricevitore.borrow_and_update().is_none() {
                return;
            }
            if ricevitore.changed().await.is_err() {
                core::future::pending::<()>().await;
            }
        }
    }
}

/// Il ciclo che tiene aggiornato lo stato.
async fn sorveglia(sentinella: Arc<Sentinella>, logind: Logind) {
    let nascite = logind.manager.receive_signal("SessionNew").await;
    let morti = logind.manager.receive_signal("SessionRemoved").await;
    let (Ok(mut nascite), Ok(mut morti)) = (nascite, morti) else {
        warn!("segnali di logind non sottoscritti: resta il solo ripasso periodico");
        ripassa_e_basta(&sentinella, &logind).await;
        return;
    };

    loop {
        // Il ripasso e i segnali fanno la stessa cosa; i segnali la fanno
        // presto, il ripasso la fa comunque.
        tokio::select! {
            _ = nascite.next() => {}
            _ = morti.next() => {}
            () = tokio::time::sleep(RIPASSO) => {}
        }
        aggiorna(&sentinella, &logind).await;
    }
}

async fn ripassa_e_basta(sentinella: &Sentinella, logind: &Logind) {
    loop {
        tokio::time::sleep(RIPASSO).await;
        aggiorna(sentinella, logind).await;
    }
}

/// Rilegge l'elenco e pubblica il risultato solo se e' cambiato.
async fn aggiorna(sentinella: &Sentinella, logind: &Logind) {
    let trovata = logind.cerca().await;
    let cambiato = *sentinella.stato.borrow() != trovata;
    if !cambiato {
        return;
    }
    match &trovata {
        Some(sessione) => warn!(%sessione, "e' comparsa una sessione grafica locale"),
        None => info!("la sessione grafica locale e' finita: si puo' entrare"),
    }
    sentinella.stato.send_replace(trovata);
}

// ---------------------------------------------------------------------------
// Il dialogo con logind
// ---------------------------------------------------------------------------

/// Quanto serve per interrogare logind, preparato una volta sola.
struct Logind {
    connessione: Connection,
    manager: Proxy<'static>,
    /// L'uid di cui ci importa: il nostro.
    uid: u32,
    /// La sessione dentro cui gira REMOTIX, da non contare mai.
    nostra: Option<String>,
}

impl Logind {
    async fn apri() -> Result<Self> {
        // Il bus di **sistema**: logind non sta su quello di sessione.
        let connessione = Connection::system()
            .await
            .context("connessione al bus di sistema")?;
        let manager = Proxy::new(
            &connessione,
            "org.freedesktop.login1",
            "/org/freedesktop/login1",
            "org.freedesktop.login1.Manager",
        )
        .await
        .context("org.freedesktop.login1 non risponde")?;

        let uid = uid_nostro().context("lettura del proprio uid")?;
        let nostra = sessione_nostra(&manager).await;
        debug!(uid, sessione = ?nostra, "sentinella pronta");
        Ok(Self {
            connessione,
            manager,
            uid,
            nostra,
        })
    }

    /// La prima sessione grafica locale del nostro utente, se c'e'.
    async fn cerca(&self) -> Option<SessioneLocale> {
        // `ListSessions` restituisce gia' uid e seat: le due condizioni che
        // scartano quasi tutto si applicano senza aprire un proxy per sessione.
        let elenco: Vec<(String, u32, String, String, OwnedObjectPath)> =
            match self.manager.call("ListSessions", &()).await {
                Ok(elenco) => elenco,
                Err(errore) => {
                    // Si prosegue senza la regola invece di chiudere fuori
                    // tutti: vedere la nota in testa al modulo.
                    warn!(errore = %errore, "elenco delle sessioni non ottenuto");
                    return None;
                }
            };

        for (id, uid, _utente, seat, percorso) in elenco {
            if uid != self.uid || seat.is_empty() {
                continue;
            }
            if self.nostra.as_deref() == Some(id.as_str()) {
                continue;
            }
            if let Some(tipo) = self.tipo_grafico(&percorso).await {
                return Some(SessioneLocale { id, tipo, seat });
            }
        }
        None
    }

    /// Il tipo della sessione, se e' grafica, dell'utente, e non sta chiudendo.
    async fn tipo_grafico(&self, percorso: &OwnedObjectPath) -> Option<String> {
        let sessione = Proxy::new(
            &self.connessione,
            "org.freedesktop.login1",
            percorso.clone(),
            "org.freedesktop.login1.Session",
        )
        .await
        .ok()?;

        // Una sessione puo' sparire fra l'elenco e queste domande: e' la
        // condizione di corsa normale di logind, non un guasto.
        let tipo: String = sessione.get_property("Type").await.ok()?;
        if !TIPI_GRAFICI.contains(&tipo.as_str()) {
            return None;
        }
        let classe: String = sessione.get_property("Class").await.ok()?;
        if classe != "user" {
            return None;
        }
        if sessione.get_property::<bool>("Remote").await.ok()? {
            return None;
        }
        // `closing` e' la sessione che se ne sta andando: contarla terrebbe
        // fuori chi si ricollega proprio mentre quella locale finisce.
        let stato: String = sessione.get_property("State").await.ok()?;
        if stato == "closing" {
            return None;
        }
        Some(tipo)
    }
}

/// L'identificatore della sessione logind dentro cui giriamo, se ce n'e' una.
///
/// Puo' non essercene: se REMOTIX viene avviato da un'unita' systemd, il suo
/// cgroup non appartiene ad alcuna sessione. In quel caso non c'e' nulla da
/// escludere, e va bene cosi'.
async fn sessione_nostra(manager: &Proxy<'static>) -> Option<String> {
    let percorso: OwnedObjectPath = manager
        .call("GetSessionByPID", &(std::process::id()))
        .await
        .ok()?;
    // L'ultimo pezzo del percorso e' l'identificatore, ma lo si chiede a logind
    // invece di ricavarlo dalla stringa: il percorso e' codificato (`session_3`
    // per la sessione `3`) e la decodifica a mano sbaglierebbe sui nomi con
    // trattini.
    let percorso = percorso.as_str().to_owned();
    let sessione = Proxy::new(
        manager.connection(),
        "org.freedesktop.login1",
        percorso,
        "org.freedesktop.login1.Session",
    )
    .await
    .ok()?;
    sessione.get_property::<String>("Id").await.ok()
}

/// Il proprio uid, letto da `/proc`.
///
/// Si legge da li' per non aggiungere una dipendenza da `libc` per una riga
/// sola. La quarta parola di `Uid:` e' quello effettivo — che e' l'uid con cui
/// logind ci vede — ma qui il processo non cambia identita', quindi il primo
/// basterebbe: si prende comunque l'effettivo, per non dover ricordare la
/// differenza il giorno in cui cambiasse.
fn uid_nostro() -> Result<u32> {
    let stato = std::fs::read_to_string("/proc/self/status").context("/proc/self/status")?;
    for riga in stato.lines() {
        if let Some(valori) = riga.strip_prefix("Uid:") {
            let campi: Vec<&str> = valori.split_whitespace().collect();
            if let Some(effettivo) = campi.get(1) {
                return effettivo.parse().context("uid non numerico");
            }
        }
    }
    anyhow::bail!("nessuna riga «Uid:» in /proc/self/status")
}

File: fondamenta/remotix-c/src/sentinella.c
#include "sentinella.h"

#include <gio/gio.h>
#include <string.h>
#include <unistd.h>

#include "registro.h"

#define NOME_LOGIND "org.freedesktop.login1"
#define PERCORSO_LOGIND "/org/freedesktop/login1"
#define IFACE_MANAGER "org.freedesktop.login1.Manager"
#define IFACE_SESSIONE "org.freedesktop.login1.Session"

/* Ogni quanto si ripassa l'elenco, oltre ai segnali. */
#define RIPASSO_MS 2000
#define ATTESA_MS 5000

struct Sentinella
{
	GThread *thread;
	GMainContext *contesto;
	GMainLoop *ciclo;

	SentinellaCambio su_cambio;
	gpointer dati;

	GDBusConnection *bus;
	guint32 uid;
	char *nostra; /* la sessione dentro cui giriamo, da non contare mai */

	/* Lo stato pubblicato, letto da altri thread. */
	GMutex lucchetto;
	gboolean presente;
	char descrizione[128];
};

static const char *TIPI_GRAFICI[] = { "wayland", "x11", "mir", NULL };

static GVariant *chiama(Sentinella *s, const char *percorso, const char *interfaccia,
                        const char *metodo, GVariant *argomenti, const GVariantType *tipo)
{
	return g_dbus_connection_call_sync(s->bus, NOME_LOGIND, percorso, interfaccia, metodo,
	                                   argomenti, tipo, G_DBUS_CALL_FLAGS_NONE, ATTESA_MS, NULL,
	                                   NULL);
}

/*
 * Il tipo della sessione, se e' grafica, dell'utente, e non sta chiudendo.
 *
 * Una sessione puo' sparire fra l'elenco e queste domande: e' la condizione di
 * corsa normale di logind, non un guasto — per questo gli errori qui si
 * ingoiano invece di essere segnalati.
 */
static char *tipo_se_grafica_locale(Sentinella *s, const char *percorso)
{
	g_autoptr(GVariant) risposta =
	    chiama(s, percorso, "org.freedesktop.DBus.Properties", "GetAll",
	           g_variant_new("(s)", IFACE_SESSIONE), G_VARIANT_TYPE("(a{sv})"));
	g_autoptr(GVariant) proprieta = NULL;
	g_autoptr(GVariant) v_tipo = NULL;
	g_autoptr(GVariant) v_classe = NULL;
	g_autoptr(GVariant) v_remota = NULL;
	g_autoptr(GVariant) v_stato = NULL;
	const char *tipo;
	gboolean grafica = FALSE;

	if (!risposta)
		return NULL;
	proprieta = g_variant_get_child_value(risposta, 0);

	v_tipo = g_variant_lookup_value(proprieta, "Type", G_VARIANT_TYPE_STRING);
	v_classe = g_variant_lookup_value(proprieta, "Class", G_VARIANT_TYPE_STRING);
	v_remota = g_variant_lookup_value(proprieta, "Remote", G_VARIANT_TYPE_BOOLEAN);
	v_stato = g_variant_lookup_value(proprieta, "State", G_VARIANT_TYPE_STRING);
	if (!v_tipo || !v_classe || !v_remota)
		return NULL;

	tipo = g_variant_get_string(v_tipo, NULL);
	for (gsize i = 0; TIPI_GRAFICI[i]; i++)
		if (g_strcmp0(tipo, TIPI_GRAFICI[i]) == 0)
			grafica = TRUE;
	if (!grafica)
		return NULL;
	if (g_strcmp0(g_variant_get_string(v_classe, NULL), "user") != 0)
		return NULL;
	if (g_variant_get_boolean(v_remota))
		return NULL;
	/* `closing` e' la sessione che se ne sta andando: contarla terrebbe fuori
	 * chi si ricollega proprio mentre quella locale finisce. */
	if (v_stato && g_strcmp0(g_variant_get_string(v_stato, NULL), "closing") == 0)
		return NULL;

	return g_strdup(tipo);
}

/* Cerca la prima sessione grafica locale del nostro utente. */
static char *cerca(Sentinella *s)
{
	g_autoptr(GVariant) risposta = chiama(s, PERCORSO_LOGIND, IFACE_MANAGER, "ListSessions", NULL,
	                                      G_VARIANT_TYPE("(a(susso))"));
	g_autoptr(GVariantIter) elenco = NULL;
	const char *id, *utente, *seat, *percorso;
	guint32 uid;

	if (!risposta)
	{
		/* Si prosegue senza la regola invece di chiudere fuori tutti. */
		avviso("elenco delle sessioni non ottenuto da logind");
		return NULL;
	}

	g_variant_get(risposta, "(a(susso))", &elenco);
	while (g_variant_iter_next(elenco, "(&su&s&s&o)", &id, &uid, &utente, &seat, &percorso))
	{
		g_autofree char *tipo = NULL;

		/* Le due condizioni che scartano quasi tutto sono gia' nell'elenco: si
		 * applicano senza aprire nulla. */
		if (uid != s->uid || !seat || !*seat)
			continue;
		if (s->nostra && g_strcmp0(id, s->nostra) == 0)
			continue;

		tipo = tipo_se_grafica_locale(s, percorso);
		if (tipo)
			return g_strdup_printf("sessione %s (%s su %s)", id, tipo, seat);
	}
	return NULL;
}

static void pubblica(Sentinella *s, char *trovata)
{
	gboolean cambiato;
	gboolean presente = trovata != NULL;

	g_mutex_lock(&s->lucchetto);
	cambiato = presente != s->presente ||
	           (presente && g_strcmp0(trovata, s->descrizione) != 0);
	if (cambiato)
	{
		s->presente = presente;
		g_strlcpy(s->descrizione, trovata ? trovata : "", sizeof s->descrizione);
	}
	g_mutex_unlock(&s->lucchetto);

	if (!cambiato)
	{
		g_free(trovata);
		return;
	}

	if (presente)
		avviso("e' comparsa una sessione grafica locale: %s", trovata);
	else
		informazione("la sessione grafica locale e' finita: si puo' entrare");

	if (s->su_cambio)
		s->su_cambio(presente, trovata, s->dati);
	g_free(trovata);
}

static gboolean ripassa(gpointer dati)
{
	Sentinella *s = dati;

	pubblica(s, cerca(s));
	return G_SOURCE_CONTINUE;
}

static void su_segnale(GDBusConnection *bus, const char *mittente, const char *percorso,
                       const char *interfaccia, const char *segnale, GVariant *parametri,
                       gpointer dati)
{
	/* Il segnale e il ripasso fanno la stessa cosa: il segnale la fa presto, il
	 * ripasso la fa comunque. */
	ripassa(dati);
}

/* L'identificatore della sessione logind dentro cui giriamo, se ce n'e' una. */
static char *sessione_nostra(Sentinella *s)
{
	g_autoptr(GVariant) risposta =
	    chiama(s, PERCORSO_LOGIND, IFACE_MANAGER, "GetSessionByPID",
	           g_variant_new("(u)", (guint32) getpid()), G_VARIANT_TYPE("(o)"));
	g_autofree char *percorso = NULL;
	g_autoptr(GVariant) proprieta = NULL;
	g_autoptr(GVariant) valore = NULL;

	if (!risposta)
		return NULL; /* avviati da un'unita' systemd: non c'e' nulla da escludere */
	g_variant_get(risposta, "(o)", &percorso);

	proprieta = chiama(s, percorso, "org.freedesktop.DBus.Properties", "Get",
	                   g_variant_new("(ss)", IFACE_SESSIONE, "Id"), G_VARIANT_TYPE("(v)"));
	if (!proprieta)
		return NULL;
	g_variant_get(proprieta, "(v)", &valore);
	return g_variant_dup_string(valore, NULL);
}

static gpointer thread_sentinella(gpointer dati)
{
	Sentinella *s = dati;

	g_main_context_push_thread_default(s->contesto);
	g_main_loop_run(s->ciclo);
	g_main_context_pop_thread_default(s->contesto);
	return NULL;
}

Sentinella *sentinella_avvia(SentinellaCambio su_cambio, gpointer dati)
{
	Sentinella *s = g_new0(Sentinella, 1);
	g_autoptr(GError) sbaglio = NULL;
	char *iniziale;
	GSource *tic;

	s->su_cambio = su_cambio;
	s->dati = dati;
	s->uid = (guint32) geteuid();
	g_mutex_init(&s->lucchetto);
	s->contesto = g_main_context_new();
	s->ciclo = g_main_loop_new(s->contesto, FALSE);

	g_main_context_push_thread_default(s->contesto);

	/* Il bus di SISTEMA: logind non sta su quello di sessione. */
	s->bus = g_bus_get_sync(G_BUS_TYPE_SYSTEM, NULL, &sbaglio);
	if (!s->bus)
	{
		/* Non ci si nasconde dietro un `diagnostica`: significa che una delle
		 * regole di §3.4 non e' in vigore, e chi legge il registro deve poterlo
		 * sapere senza andarlo a cercare. */
		avviso("logind non raggiungibile (%s): la regola della sessione locale NON e' applicata",
		       sbaglio->message);
		g_main_context_pop_thread_default(s->contesto);
		return s;
	}

	s->nostra = sessione_nostra(s);
	diagnostica("sentinella pronta: uid %u, la nostra sessione e' «%s»", s->uid,
	            s->nostra ?: "nessuna");

	/* Il primo controllo si fa PRIMA di tornare: se la macchina ha gia' una
	 * sessione grafica locale, non deve esistere una finestra iniziale in cui
	 * si entra lo stesso. */
	iniziale = cerca(s);
	if (iniziale)
	{
		s->presente = TRUE;
		g_strlcpy(s->descrizione, iniziale, sizeof s->descrizione);
		avviso("c'e' gia' una sessione grafica locale (%s): le connessioni saranno rifiutate",
		       iniziale);
		g_free(iniziale);
	}
	else
	{
		informazione("nessuna sessione grafica locale: si puo' entrare");
	}

	g_dbus_connection_signal_subscribe(s->bus, NOME_LOGIND, IFACE_MANAGER, "SessionNew",
	                                   PERCORSO_LOGIND, NULL, G_DBUS_SIGNAL_FLAGS_NONE,
	                                   su_segnale, s, NULL);
	g_dbus_connection_signal_subscribe(s->bus, NOME_LOGIND, IFACE_MANAGER, "SessionRemoved",
	                                   PERCORSO_LOGIND, NULL, G_DBUS_SIGNAL_FLAGS_NONE,
	                                   su_segnale, s, NULL);

	tic = g_timeout_source_new(RIPASSO_MS);
	g_source_set_callback(tic, ripassa, s, NULL);
	g_source_attach(tic, s->contesto);
	g_source_unref(tic);

	g_main_context_pop_thread_default(s->contesto);

	s->thread = g_thread_new("remotix-sentinella", thread_sentinella, s);
	return s;
}

gboolean sentinella_locale_presente(Sentinella *sentinella, char *descrizione, gsize quanto)
{
	gboolean presente;

	if (!sentinella)
		return FALSE;
	g_mutex_lock(&sentinella->lucchetto);
	presente = sentinella->presente;
	if (descrizione && quanto)
		g_strlcpy(descrizione, sentinella->descrizione, quanto);
	g_mutex_unlock(&sentinella->lucchetto);
	return presente;
}

void sentinella_ferma(Sentinella *sentinella)
{
	if (!sentinella)
		return;
	if (sentinella->ciclo)
		g_main_loop_quit(sentinella->ciclo);
	if (sentinella->thread)
		g_thread_join(sentinella->thread);
	g_clear_pointer(&sentinella->ciclo, g_main_loop_unref);
	g_clear_pointer(&sentinella->contesto, g_main_context_unref);
	g_clear_object(&sentinella->bus);
	g_mutex_clear(&sentinella->lucchetto);
	g_free(sentinella->nostra);
	g_free(sentinella);
}
